File: src/notebook_builder/workshop_template.java
// %% [markdown] lang="en" tags=["slide"]
//
// ## Workshop: Employees
//
// In the following we will implement a class hierarchy for employees of a company:
//
// - Employees can be either workers or managers
// - Each employee of the company has a name, a personnel number and a
//   base salary
// - For each worker, the accumulated overtime and the hourly wage
//   are stored in attributes.
// - A worker's salary is calculated as 13/12 times the
//   base salary plus overtime pay
// - Each manager has an individual bonus
// - A manager's salary is calculated as 13/12 times the
//   base salary plus bonus
//
// Implement Java classes `Employee`, `Worker` and `Manager` with
// appropriate attributes and a method `salary()` that calculates the salary.


// %% tags=["subslide"]
public class Employee {
    String name;
    String id;
    double baseSalary;

    Employee(String name, String id, double baseSalary) {
        this.name = name;
        this.id = id;
        this.baseSalary = baseSalary;
    }

    double salary() {
        return 13.0 / 12.0 * baseSalary;
    }
}

// %% 
public class Worker extends Employee {
    double overtime = 0.0;
    double hourlyWage = 0.0;
    
    Worker(String name, String id, double baseSalary, double overtime, double hourlyWage) {
        super(name, id, baseSalary);
        this.overtime = overtime;
        this.hourlyWage = hourlyWage;
    }
    
    double salary() {
        return super.salary() + this.overtime * this.hourlyWage;
    }
}

// %% 
public class Manager extends Employee {
    double bonus;

    Manager(String name, String id, double baseSalary, double bonus) {
        super(name, id, baseSalary);
        this.bonus = bonus;
    }

    double salary() {
        return super.salary() + this.bonus;
    }
}

// %% [markdown]
// Create a worker named Hans, personnel number 123, a base salary of
// 36000.0 Euros, who worked 3.5 hours of overtime at 40.0 euros each.
// Print out the salary.

// %%
Worker a = new Worker("Hans", "123", 36000.0, 3.5, 40.0);
System.out.println(a.salary());

// %% [markdown]
// Assert statements to test the functionality of the class `Worker`.

// %%
a.name.equals("Hans");
a.id.equals("123");
a.baseSalary == 36000.0;
a.overtime == 3.5;
a.hourlyWage == 40.0;
a.salary() == 39140.0;

// %% [markdown] lang="en"
// Create a manager named Sepp, personnel number 834, who is a
// base salary of 60000.0 euros and a bonus of 30000.0 euros. Print out
// the salary.

// %%
Manager m = new Manager("Sepp", "834", 60000.0, 30000.0);
System.out.println(m.salary());

// %% [markdown] lang="en"
// Test the functionality of the class `Manager`.

// %%
m.salary() == 95000.0;

File: src/notebook_builder/workshop_template.py
# %% [markdown] lang="en" tags=["slide"]
#
# ## Workshop: Employees
#
# In the following we will implement a class hierarchy for employees of a company:
#
# - Employees can be either workers or managers
# - Each employee of the company has a name, a personnel number and a
#   base salary
# - For each worker, the accumulated overtime and the hourly wage
#   are stored in attributes.
# - A worker's salary is calculated as 13/12 times the
#   base salary plus overtime pay
# - Each manager has an individual bonus
# - A manager's salary is calculated as 13/12 times the
#   base salary plus bonus
#
# Implement Python classes `Employee`, `Worker` and `Manager` with
# appropriate attributes and a method `salary()` that calculates the salary.


# %% tags=["alt"]
class Employee:
    def __init__(self, name, pers_nr, base_salary):
        self.name = name
        self.pers_nr = pers_nr
        self.base_salary = base_salary

    def salary(self):
        return 13 / 12 * self.base_salary


# %% tags=["alt"]
class Worker(Employee):
    def __init__(self, name, pers_nr, base_salary, overtime, hourly_wage):
        super().__init__(name, pers_nr, base_salary)
        self.overtime = overtime
        self.hourly_wage = hourly_wage

    def __repr__(self):
        return (
            f"Worker({self.name!r}, {self.pers_nr!r}, {self.base_salary}, "
            f"{self.overtime}, {self.hourly_wage})"
        )

    def salary(self):
        return super().salary() + self.overtime * self.hourly_wage


# %% tags=["alt"]
class Manager(Employee):
    def __init__(self, name, pers_nr, base_salary, bonus):
        super().__init__(name, pers_nr, base_salary)
        self.bonus = bonus

    def __repr__(self):
        return (
            f"Manager({self.name!r}, {self.pers_nr!r}, {self.base_salary}, "
            f"{self.bonus})"
        )

    def salary(self):
        return super().salary() + self.bonus


# %% tags=["alt"]
a = Worker("Hans", 123, 36_000, 3.5, 40)

# %%
assert a.salary() == 39_140.0

# %% tags=["alt"]
m = Manager("Sepp", 834, 60_000, 30_000)

# %%
assert m.salary() == 95_000.0
